### PR TITLE
CI: Start testing Python 3.12

### PR DESCRIPTION
**Pull request by EwoutH**
Add Python 3.12 to the Ubuntu job matrix to start testing the betas in the CI.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -39,6 +39,8 @@ jobs:
       with:
         path: ~/.cache/pip
         key: pip-test-${{ matrix.python-version }}-${{ matrix.os }}
+    - name: Install setuptools
+      run: pip install setuptools
```
Comment (agronholm) [inline .github/workflows/test.yml:43]: If setuptools is a crucial testing dependency, maybe it should instead be added to test dependencies in `pyproject.toml`?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -14,7 +14,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest]
-        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11", "pypy-3.8", "pypy-3.9"]
+        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11", "3.12-dev", "pypy-3.8", "pypy-3.9"]
```
Comment (agronholm) [inline .github/workflows/test.yml:17]: It would be preferable to use a combination of `"3.12"` and then `allow-prereleases: true` below.

**Review by agronholm — CHANGES_REQUESTED** [review]

**Comment by EwoutH** [comment]
@agronholm Thanks for reviewing! I implemented your suggested changes.

**Comment by henryiii** [comment]
I'll add a test to #529 once this goes in.

**Comment by agronholm** [comment]
Thanks!

**Pull request merged**